pkg/networkserver/grpc_asns.go

```diff
@@ -65,32 +71,36 @@ func (ns *NetworkServer) LinkApplication(link ttnpb.AsNs_LinkApplicationServer)
 	ns.applicationServersMu.Lock()
 	cl, ok := ns.applicationServers[uid]
 	ns.applicationServers[uid] = ws
+	ns.applicationServersMu.Unlock()
 	if ok {
+		logger.Debug("Close existing link")
 		if err := cl.Close(); err != nil {
-			ns.applicationServersMu.Unlock()
-			logger.WithError(err).Warn("Failed to link application")
-			return err
+			logger.WithError(err).Warn("Failed to close existing link")
 		}
 	}
-	ns.applicationServersMu.Unlock()
 
 	logger.Debug("Linked application")
 
 	events.Publish(evtBeginApplicationLink(ctx, ids, nil))
 	defer events.Publish(evtEndApplicationLink(ctx, ids, err))
 
-	select {
-	case <-ctx.Done():
-		err := ctx.Err()
+	defer func() {
 		ns.applicationServersMu.Lock()
-		cl, ok := ns.applicationServers[uid]
+		cl, ok = ns.applicationServers[uid]
 		if ok && cl == ws {
 			delete(ns.applicationServers, uid)
 		}
 		ns.applicationServersMu.Unlock()
+	}()
+
+	select {
+	case <-ctx.Done():
+		err = ctx.Err()
+		logger.WithError(err).Debug("Context done - close link")
 		return err
-	case <-ws.closeCh:
-		return errDuplicateSubscription
+	case ws.closeCh <- struct{}{}:
+		logger.Debug("Close() is called, close link")
```

Comment:
Inconsistent with debug log above, which uses `- close link`

Reply:
Even though it's Debug I would not print code